src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs

```diff
@@ -0,0 +1,352 @@
+            }
+
+            public sealed override Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+                => _codeStyleProvider.ComputeRefactoringsAsync(context);
+        }
+
+        #endregion
+
+        #region fixing
+
+        public abstract class CodeStyle2CodeFixProvider : SyntaxEditorBasedCodeFixProvider
+        {
+            public readonly TCodeStyleProvider _codeStyleProvider;
+
+            protected CodeStyle2CodeFixProvider()
+            {
+                _codeStyleProvider = new TCodeStyleProvider();
+                FixableDiagnosticIds = ImmutableArray.Create(_codeStyleProvider._descriptorId);
+            }
+
+            public sealed override ImmutableArray<string> FixableDiagnosticIds { get; }
+
+            public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+                => _codeStyleProvider.RegisterCodeFixesAsync(this, context);
+
+            protected sealed override Task FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken)
+                => _codeStyleProvider.FixAllAsync(document, diagnostics, editor, cancellationToken);
+        }
+
+        private async Task RegisterCodeFixesAsync(
+            CodeStyle2CodeFixProvider fixProvider,
+            CodeFixContext context)
+        {
+            _fixProvider = fixProvider;
+
+            var document = context.Document;
+            var diagnostic = context.Diagnostics[0];
+            var cancellationToken = context.CancellationToken;
+
+            var codeFixes = await ComputeCodeActionsAsync(
+                document, diagnostic, cancellationToken).ConfigureAwait(false);
+            context.RegisterFixes(codeFixes, context.Diagnostics);
+        }
+
+        protected abstract Task<ImmutableArray<CodeAction>> ComputeCodeActionsAsync(
+            Document document, Diagnostic diagnostic, CancellationToken cancellationToken);
+
+        protected abstract Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        protected Task<Document> FixAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+            => _fixProvider.FixAsync(document, diagnostic, cancellationToken);
+
+        #endregion
+    }
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class ExampleDiagnosticAnalyzer : ExampleCodeStyleProvider.CodeStyle2DiagnosticAnalyzer
+    {
+    }
+
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = "Example code refactoring provider")]
+    class ExampleCodeRefactoringProvider : ExampleCodeStyleProvider.CodeStyle2CodeRefactoringProvider
+    {
+    }
+
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "Example code fix provider")]
+    class ExampleCodeFixProvider : ExampleCodeStyleProvider.CodeStyle2CodeFixProvider
+    {
+    }
```

Comment:
These are the three classes a feature would still have to export. Note they are very intentionally empty. They do not need to do anything, and literally exist solely so we can add the export attributes somewhere so these can be found by the relevant subsystems.

in an ideal world we could just put these attributes on the CodeStyleProvider. however, since each of these attributes expects a particular inheritance hierarchy, and because they all use classes, there is no way to make CodeStyleProvider extend all three of these classes.

Alternatively, we could just have something like ExportCodeStyleProvider, and have the relevant subsystems know how to import that. Then, the CodeStyleProvider would expose the 3 separate providers from itself. That would be trivial to do for CodeRefactoringProvider/CodeFixProvider. however, it's unclear to me how we would fit that into the DiagnosticAnalyzer infrastructure.